zrml/hybrid-router/src/lib.rs

```diff
@@ -0,0 +1,630 @@
+                        );
+                        ensure!(order_price >= price_limit, Error::<T>::OrderPriceBelowMinPrice);
+                    }
+                }
+
+                remaining = Self::maybe_fill_from_amm(
+                    tx_type,
+                    who,
+                    market_id,
+                    asset,
+                    remaining,
+                    order_price,
+                )?;
+
+                if remaining.is_zero() {
+                    break;
+                }
+
+                // `remaining` is always denominated in the `taker_asset`
+                // because this is what the order owner (maker) wants to receive
+                let (_taker_fill, maker_fill) =
+                    order.taker_and_maker_fill_from_taker_amount(remaining)?;
+                // and the `maker_partial_fill` of `fill_order` is specified in `taker_asset`
+                T::OrderBook::fill_order(who, order_id, Some(maker_fill))?;
+                // `maker_fill` is the amount the order owner (maker) wants to receive
+                remaining = remaining.checked_sub_res(&maker_fill)?;
+            }
+
+            Ok(remaining)
+        }
+
+        /// Places a limit order if the strategy is `Strategy::LimitOrder`.
+        /// If the strategy is `Strategy::ImmediateOrCancel`, an error is returned.
+        ///
+        /// # Arguments
+        ///
+        /// * `strategy` - The strategy to handle the remaining non-zero amount when the `max_price` is reached.
+        /// * `who` - The account ID of the user performing the transaction.
+        /// * `market_id` - The ID of the market.
+        /// * `maker_asset` - The asset to provide.
+        /// * `maker_amount` - The amount of the `maker_asset` to be provided.
+        /// * `taker_asset` - The asset to be received.
+        /// * `taker_amount` - The amount of the `taker_asset` to be received.
+        ///
+        /// # Returns
+        ///
+        /// An error if the strategy is `Strategy::ImmediateOrCancel`.
+        /// Otherwise, the limit order is placed.
+        fn maybe_place_limit_order(
+            strategy: Strategy,
+            who: &AccountIdOf<T>,
+            market_id: MarketIdOf<T>,
+            maker_asset: AssetOf<T>,
+            maker_amount: BalanceOf<T>,
+            taker_asset: AssetOf<T>,
+            taker_amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            match strategy {
+                Strategy::ImmediateOrCancel => {
+                    return Err(Error::<T>::CancelStrategyApplied.into());
+                }
+                Strategy::LimitOrder => {
+                    T::OrderBook::place_order(
+                        who,
+                        market_id,
+                        maker_asset,
+                        maker_amount,
+                        taker_asset,
+                        taker_amount,
+                    )?;
+                }
+            }
+
```

Comment:
We have a similar problem here where a user might end up with a very small `remaining` amount. If their strategy is `LimitOrder`, this call will then fail due to the minimum order size set in zrml-orderbook.